packet/src/pacer.rs

```diff
@@ -0,0 +1,1032 @@
+use std::collections::HashMap;
+use std::fmt;
+use std::time::{Duration, Instant};
+
+use rtp::{Bitrate, DataSize};
+
+const MAX_BITRATE: Bitrate = Bitrate::gbps(10);
+const MAX_MEDIA_DEBT_IN_TIME: Duration = Duration::from_millis(500);
+
+pub enum PacerImpl {
+    Null(NullPacer),
+    LeakyBucket(LeakyBucketPacer),
+}
+
+impl Pacer for PacerImpl {
+    fn set_target_bitrate(&mut self, target_bitrate: Bitrate) {
+        match self {
+            PacerImpl::Null(v) => v.set_target_bitrate(target_bitrate),
+            PacerImpl::LeakyBucket(v) => v.set_target_bitrate(target_bitrate),
+        }
+    }
+
+    fn poll_timeout(&self) -> Option<Instant> {
+        match self {
+            PacerImpl::Null(v) => v.poll_timeout(),
+            PacerImpl::LeakyBucket(v) => v.poll_timeout(),
+        }
+    }
+
+    fn handle_timeout(&mut self, now: Instant, iter: impl Iterator<Item = QueueState>) {
+        match self {
+            PacerImpl::Null(v) => v.handle_timeout(now, iter),
+            PacerImpl::LeakyBucket(v) => v.handle_timeout(now, iter),
+        }
+    }
+
+    fn poll_action(&mut self) -> PollOutcome {
+        match self {
+            PacerImpl::Null(v) => v.poll_action(),
+            PacerImpl::LeakyBucket(v) => v.poll_action(),
+        }
+    }
+
+    fn register_send(&mut self, now: Instant, packet_size: DataSize, from: QueueId) {
+        match self {
+            PacerImpl::Null(v) => v.register_send(now, packet_size, from),
+            PacerImpl::LeakyBucket(v) => v.register_send(now, packet_size, from),
+        }
+    }
+}
+
+/// A packet Pacer.
+///
+/// The pacer is responsible for ensuring correct pacing of packets onto the network at a given
+/// bitrate.
+pub trait Pacer {
+    /// Set the target bitrate to send at. This can be exceeded if required to drain excessively
+    /// long packet queues.
```

Comment:
I think whether or not the given rate should be exceeded should be configurable.  For example, libwebrtc has the PacingController::drain_large_queues_ to control that.  

And we should probably comment that if you enable sending above the given rate, that's really only intended to handle bursts (like key frames).  If you send over a long period above the given rate and enable going above it, you can of defeat the purpose of having a given rate in the first place.